serde-generate/src/rust.rs

```diff
@@ -0,0 +1,204 @@
+// Copyright (c) Facebook, Inc. and its affiliates
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+use crate::analyzer;
+use serde_reflection::{ContainerFormat, Format, Named, Registry, VariantFormat};
+use std::collections::{BTreeMap, HashSet};
+use std::io::{Result, Write};
+
+pub fn output(
+    out: &mut dyn Write,
+    with_derive_macros: bool,
+    registry: &Registry,
+) -> std::result::Result<(), Box<dyn std::error::Error>> {
+    let dependencies = analyzer::get_dependency_map(registry)?;
+    let entries = analyzer::best_effort_topological_sort(&dependencies)?;
+
+    output_preambule(out, with_derive_macros)?;
+    let mut known_sizes = HashSet::new();
+    for name in entries {
+        let format = &registry[name];
+        output_container(out, with_derive_macros, name, format, &known_sizes)?;
+        known_sizes.insert(name);
+    }
+    Ok(())
+}
+
+fn output_preambule(out: &mut dyn Write, with_derive_macros: bool) -> Result<()> {
+    let serde_imports = if with_derive_macros {
+        r#"
+use serde::{Serialize, Deserialize};
+use serde_bytes::ByteBuf;"#
+    } else {
+        // for testing
+        "type ByteBuf = Vec<u8>;"
+    };
+    writeln!(
+        out,
+        r#"
+#![allow(unused_imports)]{}
+use std::collections::BTreeMap;
```

Comment:
Maybe you could use the type directly with a fully qualified path instead of having to allow unused imports?

Reply:
Good idea. Let me simplify this.